Ladybird/LocationEdit.cpp

```diff
@@ -14,4 +17,74 @@ LocationEdit::LocationEdit(Tab* tab)
     connect(this, &QLineEdit::returnPressed, this, [&] {
         clearFocus();
     });
+
+    connect(this, &QLineEdit::textChanged, this, [&] {
+        highlightLocation();
+    });
+}
+
+void LocationEdit::focusInEvent(QFocusEvent* event)
+{
+    QLineEdit::focusInEvent(event);
+    highlightLocation();
+}
+
+void LocationEdit::focusOutEvent(QFocusEvent* event)
+{
+    QLineEdit::focusOutEvent(event);
+    highlightLocation();
+}
+
+void LocationEdit::highlightLocation()
+{
+    AK::URL url = AK::URL::create_with_url_or_path(text().toStdString().c_str());
+
+    QList<QInputMethodEvent::Attribute> attributes;
+    if (url.is_valid() && !hasFocus()) {
+        if (url.scheme() == "http" || url.scheme() == "https" || url.scheme() == "gemini") {
+            int host_start = (url.scheme().length() + 3) - cursorPosition();
+            auto host_length = url.host().length();
+
+            if (auto host_parts = url.host().split('.'); host_parts.size() > 2) {
+                host_length = 0;
+                for (auto i = host_parts.size() - 2; i < host_parts.size(); i++) {
```

Comment:
Assuming I'm reading this right and this fades out all but the last 2 parts of the domain, it'll highlight URLs like `https://www.bbc.co.uk/` incorrectly, since that has 3 parts. It's common in the UK to have URLs with two top-level parts like that. (Probably some other countries too.)

Reply:
I was linked to this resource [publicsuffix.org](https://publicsuffix.org/list/public_suffix_list.dat). It seems like exactly what we would want here, however, I'm not sure how I would even start to use it.

Reply:
Oh nice. Yeah, that's going to be a bit more involved. What you'll want is that during the build, you download that file, parse it, and generate some C++ functions that can tell you which parts of the URL to highlight.

We do some similar code-generation already for LibUnicode and some other libraries. Tim Flynn is a good person to ask about it. But if you're not comfortable doing that right now, just add a FIXME about it and someone else can get to it. :^)

Reply:
FIXME has been added, I've removed the partial host highlighting for now. Couldn't for the life of me figure out how to add a new generator